### PR TITLE
[MM-56183] Handle bot clients ws reconnects

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

During a transcribing job the ws connection briefly dropped, the client immediately reconnected without issues but our server side logic didn't account for this to happen. Upon stopping the job we'd be sending the fake `call_end` event to the outdated connection ID which would cause the client to never receive it and the job to continue indefinitely.

PR addresses this by handling this special case without introducing breaking changes. Upon reconnect we update the bot connection id in the state.

A better approach would be to send an explicit event with the job ID included in the payload so that there's no confusion no matter how many times the bot client reconnected. I am planning to do this in https://mattermost.atlassian.net/browse/MM-56187.

Will need to do some manual testing on HA environment prior to merging.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-56183

### CODE REVIEW

File: server/plugin.go [hunk 1]
```diff
@@ -65,7 +65,8 @@ type Plugin struct {
 
 	// A map of callID -> *cluster.Mutex to guarantee atomicity of call state
 	// operations.
-	callsClusterLocks map[string]*cluster.Mutex
+	callsClusterLocks    map[string]*cluster.Mutex
+	callsClusterLocksMut sync.RWMutex
```
Comment (cpoile) [inline server/plugin.go:69]: We're using a mut for a map of mutexes... 😂

**Review by cpoile — APPROVED** [review]

File: server/plugin.go (hunk 1 above)
Comment (streamer45) [reply server/plugin.go:69]: indeed, better than deadlocking which was the alternative :)

**Comment by streamer45** [comment]
@cpoile Asking for a second review as last commit was significant :sweat_smile: 

While not in love with mocked tests, I found a bug while doing manual testing that a properly mocked unit test would have caught. So added some boilerplate to make it work. Hopefully it will also make it easier to test more stuff going forward and bump our repo stats as a bonus :)

File: server/websocket_test.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+			}).Once()
+
+			state.Call.Transcription.BotConnID = "newTranscribingBotConnID"
+			expectedStateJSON, err := json.Marshal(state)
+			require.NoError(t, err)
+
+			mockAPI.On("KVSet", "channelID", expectedStateJSON).Return(nil).Once()
+
+			err = p.handleBotWSReconnect("newTranscribingBotConnID", "prevTranscribingBotConnID", "originalConnID", channelID)
+			require.NoError(t, err)
```
Comment (cpoile) [inline server/websocket_test.go:225]: Could you explain a bit about what this is testing? It's a little unclear to me looking from the outside.

**Review by cpoile — COMMENTED** [review]
Just a clarifying question

File: server/websocket_test.go (hunk 2 above)
Comment (streamer45) [reply server/websocket_test.go:225]: The most important part is the `KVSet` mocked call. We are essentially asserting the state mutation is what's expected given the initial `state` and the `handleBotWSReconnect` method's inputs. Let me add a couple of comments.

Comment (streamer45) [reply server/websocket_test.go]: Done

**Review by cpoile — APPROVED** [review]
Ok, that clearer, thank you!

**Pull request merged**